File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if not intervals:
            return [newInterval]
        ans = []
        is_inserted = False
        for interval in intervals:
            if is_inserted:
                ans.append(interval)
            # [left <= right < new_left]
            elif interval[1] < newInterval[0]:
                ans.append(interval)
            # [left <= new_right or right]
            elif interval[0] <= newInterval[1]:
                newInterval[0] = min(newInterval[0], interval[0])
                newInterval[1] = max(newInterval[1], interval[1])
            # [new_right < left]
            else:
                is_inserted = True
                ans.append(newInterval)
                ans.append(interval)
            # print(ans, newInterval)
        if not is_inserted:
            ans.append(newInterval)
        return ans

File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        int start=newInterval[0], end=newInterval[1];
        int i=0, n=intervals.size();
        vector<vector<int>> ans;
        while (i<n and start>intervals[i][0]) ans.push_back(intervals[i++]);
        vector<int> cur(2);
        if (ans.empty() or ans.back()[1]<start) ans.push_back(newInterval);
        else ans.back()[1]=max(ans.back()[1],end);
        while (i<n){
            cur=intervals[i++];
            if (ans.back()[1]<cur[0]) ans.push_back(cur);
            else ans.back()[1]=max(ans.back()[1],cur[1]);
        }
        return ans;
    }
};